diff --git a/ext/opcache/jit/zend_jit_trace.c b/ext/opcache/jit/zend_jit_trace.c
index f89d6f8a40..075acf5e6f 100644
--- a/ext/opcache/jit/zend_jit_trace.c
+++ b/ext/opcache/jit/zend_jit_trace.c
@@ -6377,7 +6377,7 @@ static const void *zend_jit_trace(zend_jit_trace_rec *trace_buffer, uint32_t par
 					}
 					SET_STACK_TYPE(stack, EX_VAR_TO_NUM(opline->op2.var), type,
 						(gen_handler || type == IS_UNKNOWN || !ra ||
-							(!ra[ssa_op->op2_def] && !ssa->vars[ssa_op->op2_def].no_val)));
+							(!ra[ssa_op->op2_def] /*&& !ssa->vars[ssa_op->op2_def].no_val*/)));
 					if (type != IS_UNKNOWN) {
 						ssa->var_info[ssa_op->op2_def].type &= ~MAY_BE_GUARD;
 						if (ra && ra[ssa_op->op2_def]) {
diff --git a/ext/opcache/tests/jit/assign_054.phpt b/ext/opcache/tests/jit/assign_054.phpt
new file mode 100644
index 0000000000..a4c4706b7c
--- /dev/null
+++ b/ext/opcache/tests/jit/assign_054.phpt
@@ -0,0 +1,24 @@
+--TEST--
+JIT ASSIGN: missing type store
+--INI--
+opcache.enable=1
+opcache.enable_cli=1
+opcache.file_update_protection=0
+opcache.jit_buffer_size=1M
+opcache.protect_memory=1
+--FILE--
+<?php
+function foo($a) {
+    $b = $a = $a + $b & $a += $a;
+    $b = $a = $a + $b & $b & $b = $a = $a + $b = $a = $a + $b = $a += $a;
+    $b = !$a = $a + $b & $b & $b = $b = $a = $a + $b & $a += $a;
+	$a + $b & $b & $b = $a = $a + $b = $a = $a + $b = $a += $a;
+}
+
+@foo(39087589046889428661);
+@foo(390875890468877606478);
+@foo(390875890468877606478);
+?>
+DONE
+--EXPECT--
+DONE
