diff --git a/ext/opcache/jit/zend_jit_helpers.c b/ext/opcache/jit/zend_jit_helpers.c
index f1a24ea856..a36cccb401 100644
--- a/ext/opcache/jit/zend_jit_helpers.c
+++ b/ext/opcache/jit/zend_jit_helpers.c
@@ -2379,6 +2379,9 @@ static zval * ZEND_FASTCALL zend_jit_prepare_assign_dim_ref(zval *ref) {
 		if (Z_TYPE_P(val) == IS_FALSE) {
 			ZVAL_ARR(val, zend_new_array(8));
 			zend_false_to_array_deprecated();
+			if (EG(exception)) {
+				return NULL;
+			}
 		} else {
 			ZVAL_ARR(val, zend_new_array(8));
 		}
diff --git a/ext/opcache/tests/jit/assign_dim_011.phpt b/ext/opcache/tests/jit/assign_dim_011.phpt
new file mode 100644
index 0000000000..941033a8ac
--- /dev/null
+++ b/ext/opcache/tests/jit/assign_dim_011.phpt
@@ -0,0 +1,25 @@
+--TEST--
+JIT ASSIGN_DIM: 011
+--INI--
+opcache.enable=1
+opcache.enable_cli=1
+opcache.file_update_protection=0
+opcache.jit_buffer_size=1M
+--FILE--
+<?php
+$my_var = null < 
+set_error_handler(function($code, $msg) use(&$my_var) {
+	echo "Err: $msg\n";
+    $my_var[] = $my_var = 0;
+});
+try {
+    $my_var[] = "";
+} catch (Throwable $e) {
+    echo "Exception: " . $e->getMessage() . "\n";
+}
+?>
+DONE
+--EXPECT--
+Err: Automatic conversion of false to array is deprecated
+Exception: Cannot use a scalar value as an array
+DONE
