diff --git a/Zend/Optimizer/block_pass.c b/Zend/Optimizer/block_pass.c
index b176ba2704..e522f65067 100644
--- a/Zend/Optimizer/block_pass.c
+++ b/Zend/Optimizer/block_pass.c
@@ -256,6 +256,13 @@ static void zend_optimize_block(zend_basic_block *block, zend_op_array *op_array
 				}
 				break;
 
+			case ZEND_MATCH_ERROR:
+				if (opline->op1_type == IS_TMP_VAR) {
+					src = VAR_SOURCE(opline->op1);
+					VAR_SOURCE(opline->op1) = NULL;
+				}
+				break;
+
 			case ZEND_FREE:
 				if (opline->op1_type == IS_TMP_VAR) {
 					src = VAR_SOURCE(opline->op1);
diff --git a/ext/opcache/tests/opt/match_002.phpt b/ext/opcache/tests/opt/match_002.phpt
new file mode 100644
index 0000000000..ecda31ee7c
--- /dev/null
+++ b/ext/opcache/tests/opt/match_002.phpt
@@ -0,0 +1,15 @@
+--TEST--
+Match 002: memory leak because of incorrect optimization
+--INI--
+opcache.enable=1
+opcache.enable_cli=1
+opcache.optimization_level=-1
+--FILE--
+<?php
+match($y=y){};
+?>
+--EXPECTF--
+Fatal error: Uncaught Error: Undefined constant "y" in %smatch_002.php:2
+Stack trace:
+#0 {main}
+  thrown in %smatch_002.php on line 2
