diff --git a/ext/opcache/Optimizer/sccp.c b/ext/opcache/Optimizer/sccp.c
index 8163929b39..903a4122c1 100644
--- a/ext/opcache/Optimizer/sccp.c
+++ b/ext/opcache/Optimizer/sccp.c
@@ -1545,13 +1545,13 @@ static void sccp_visit_instr(scdf_ctx *scdf, zend_op *opline, zend_ssa_op *ssa_o
 
 						dup_partial_object(&zv, op1);
 						ct_eval_assign_obj(&zv, &tmp2, op2);
-						if (opline->opcode == ZEND_PRE_INC_OBJ
-								|| opline->opcode == ZEND_PRE_DEC_OBJ) {
+						if (opline->opcode == ZEND_PRE_INC_OBJ || opline->opcode == ZEND_PRE_DEC_OBJ) {
 							SET_RESULT(result, &tmp2);
-							zval_ptr_dtor_nogc(&tmp1);
 						} else {
 							SET_RESULT(result, &tmp1);
 						}
+						zval_ptr_dtor_nogc(&tmp1);
+						zval_ptr_dtor_nogc(&tmp2);
 						SET_RESULT(op1, &zv);
 						zval_ptr_dtor_nogc(&zv);
 						break;
diff --git a/ext/opcache/tests/opt/sccp_038.phpt b/ext/opcache/tests/opt/sccp_038.phpt
new file mode 100644
index 0000000000..fdf663e87c
--- /dev/null
+++ b/ext/opcache/tests/opt/sccp_038.phpt
@@ -0,0 +1,18 @@
+--TEST--
+SCCP 038: Memory leak
+--INI--
+opcache.enable=1
+opcache.enable_cli=1
+opcache.optimization_level=-1
+--FILE--
+<?php
+function foo() {
+    $obj = new stdClass;
+    $obj->$b = ~$b = $a = '##';
+    $obj->$a++;
+}
+foo();
+?>
+DONE
+--EXPECT--
+DONE
