diff --git a/Zend/tests/class_constants_007.phpt b/Zend/tests/class_constants_007.phpt
new file mode 100644
index 0000000000..d09a12e8c1
--- /dev/null
+++ b/Zend/tests/class_constants_007.phpt
@@ -0,0 +1,13 @@
+--TEST--
+Ownership of constant expression inhereted from immutable class should be transfered to class
+--FILE--
+<?php
+class A {
+    const X = ' ' . self::Y;
+    const Y = ' ';
+}
+eval('class B extends A{}');
+var_dump(B::X);
+?>
+--EXPECT--
+string(2) "  "
diff --git a/Zend/zend_inheritance.c b/Zend/zend_inheritance.c
index acff1da179..e91195bb15 100644
--- a/Zend/zend_inheritance.c
+++ b/Zend/zend_inheritance.c
@@ -1354,6 +1354,7 @@ static void do_inherit_class_constant(zend_string *name, zend_class_constant *pa
 				c = zend_arena_alloc(&CG(arena), sizeof(zend_class_constant));
 				memcpy(c, parent_const, sizeof(zend_class_constant));
 				parent_const = c;
+				Z_CONSTANT_FLAGS(c->value) |= CONST_OWNED;
 			}
 		}
 		if (ce->type & ZEND_INTERNAL_CLASS) {
